find-modified-layout-tests step should detect the tests modified in test expectations in a PR

briannafan opened the pull request:
#### e530f8ef12baeb5fa5cc60e2e44122a7445f1a3b
<pre>
find-modified-layout-tests step should detect the tests modified in test expectations in a PR
<a href="https://bugs.webkit.org/show_bug.cgi?id=271721">https://bugs.webkit.org/show_bug.cgi?id=271721</a>
<a href="https://rdar.apple.com/problem/123398588">rdar://problem/123398588</a>

Reviewed by Jonathan Bedard.

Add additional steps to check if TestExpectations is modified.
These save the output of run-webkit-tests --print-expectations before and after the PR/patch is applied.
Then these two results are compared to get any changes to test expectations.

* Tools/CISupport/ews-build/factories.py:
(Factory): Add skipBuildIfNoResult property.
(Factory.__init__): Add GetTestExpectationsBaseline() and GetUpdatedTestExpectations(). Move FindModifiedLayoutTests().
(TestFactory): Set skipBuildIfNoResult.
(iOSTestsFactory): Set findModifiedLayoutTests.
(macOSWK1Factory): Ditto.
(macOSWK2Factory): Ditto.
(GTKTestsFactory): Ditto.
(WPETestsFactory): Ditto.

* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps):

* Tools/CISupport/ews-build/steps.py:
(GetTestExpectationsBaseline): Print test expectations before applying changes.
(GetTestExpectationsBaseline.run):
(GetUpdatedTestExpectations): Print test expectations again after applying changes.
(GetUpdatedTestExpectations.run):
(FindModifiedLayoutTests): Parse diff between two expectations and check patch for LayoutTest changes.
(FindModifiedLayoutTests.run):
(FindModifiedLayoutTests.getResultSummary):
(FindModifiedLayoutTests.start): Deleted.

* Tools/CISupport/ews-build/steps_unittest.py:
(TestGetTestExpectationsBaseline): Added.
(TestGetUpdatedTestExpectations): Added.
(TestFindModifiedLayoutTests):
(TestFindModifiedLayoutTests.test_non_accessible_patch): Test for non-accessible patch/PR contents.

Canonical link: <a href="https://commits.webkit.org/277546@main">https://commits.webkit.org/277546@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dbe93426f594fa52692120639b69568f165bdb0b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47913 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27114 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50771 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50594 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43966 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50219 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32986 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24599 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38979 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48495 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24775 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41435 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20277 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22256 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5960 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44261 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52488 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22955 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19301 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46288 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/47946 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24225 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45332 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25016 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6795 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23946 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/22c295eb9e5497342f6c2831771a3f39d000bb17. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26478

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7010f0b25143164ced8219cdb39bc7a5fe210921. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26478

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bfc379c48b5d2dc7b7e29b21f6c65a94634b959c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26478

aproskuryakov commented:
Does this cover unskipping a whole directory?

I was expecting that this PR would use dry run to get a definitive list of expectations, not grepping.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1383,12 +1385,25 @@ def __init__(self, skipBuildIfNoResult=True):
+        test_expectation_changes = False
         for line in patch.splitlines():
+            # Check if changes are in a TestExpectations file
+            file_change_line = re.search(self.RE_FILE_CHANGE, line, re.IGNORECASE)
```
Have we double checked that the "patch diff" for a GitHub PR actually has the diff of each file? I seem to recall we only implemented which files were added/removed/modified, not what that modification actually was

briannafan replied:
Addressed IRL. For posterity, this is true and the implementation has been changed to look at the results of `run-webkit-tests --print-expectations`.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/afc655fcf2ecf7034adcb683d0e346184879aa32. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26478

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/324e757999aa19593a99b59aac832661f30c5d44. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26478

aproskuryakov commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1371,8 +1371,38 @@ def start(self):
+        if patch_author in ['webkit-wpt-import-bot@igalia.com']:
+            self.setCommand(self.command + ['imported/w3c/web-platform-tests'])
+
+        self.setCommand(['/bin/sh', '-c', ' '.join(self.command) + ' > base-expectations.txt'])
```
What is the reason to execute via shell, not directly?

briannafan replied:
I think it's necessary to execute via shell to pipe stdout into a text file, correct me if I'm wrong.

JonWBedard replied:
This is indeed how we achieve something like "save file on worker containing the output of a script". Although @briannafan, I would use `ShellMixin`'s `shell_command`, that way this is Windows compatible.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ebd21a9db2c286699e483371a6a7397a291d3987. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26478

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/490d3d418a13797adec1e0980dd0fadb828b91ba. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26478

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4fe4d9f3014067c742cc672e775ed5fff927370f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26478

briannafan commented:
Tested on staging:

1. Layout tests modified: [Build 155](https://ews-build.webkit-uat.org/#/builders/74/builds/155) -> Success 
2. No relevant changes: [Build 156](https://ews-build.webkit-uat.org/#/builders/74/builds/156) -> Skipped
3. Unable to access patch/PR contents: [Build 158](https://ews-build.webkit-uat.org/#/builders/74/builds/158) -> Warning

Note that the exceptions in `download-built-product` for builds 156 and 158 are expected since I forced tested patches that were not built.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1314,16 +1314,127 @@ def start(self):
+
+        patch_author = self.getProperty('patch_author')
+        if patch_author in ['webkit-wpt-import-bot@igalia.com']:
+            run_webkit_command += ['imported/w3c/web-platform-tests']
```
Don't we need to do something similar for iOS WPT testers?

JonWBedard replied:
I mean that WPT tests also only run tests from `imported/w3c/web-platform-tests`, and for iOS, one of our testers explicitly excludes WPT tests. I think that we should be matching the path arguments those testers provide, and I don't think that we are, at the moment.

briannafan replied:
Ah, got it! I missed this because it was set in `config.json` but I've now added `additionalArguments` to the command.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1314,16 +1314,127 @@ def start(self):
+        self.commands = []
+        for command in [
+            self.shell_command(run_webkit_command),
+            self.shell_command("sed -i '' 's/\].*/\]/' base-expectations.txt"),
```
What exactly are we doing with `sed` here? Also, does Windows have `sed`?

briannafan replied:
The output of `run-webkit-tests --print-expectations` includes line numbers which can change even though the actual expectation or test name didn't, so I'm removing that section and comparing on the test name and expectation only.
Good point about Windows - I will explore a solution that works on those bots as well.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/99c43f6e15ac7611d494b1605dc1251397f1c050. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26478

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/dbe93426f594fa52692120639b69568f165bdb0b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26478

briannafan committed:
find-modified-layout-tests step should detect the tests modified in test expectations in a PR
https://bugs.webkit.org/show_bug.cgi?id=271721
rdar://problem/123398588

Reviewed by Jonathan Bedard.

Add additional steps to check if TestExpectations is modified.
These save the output of run-webkit-tests --print-expectations before and after the PR/patch is applied.
Then these two results are compared to get any changes to test expectations.

* Tools/CISupport/ews-build/factories.py:
(Factory): Add skipBuildIfNoResult property.
(Factory.__init__): Add GetTestExpectationsBaseline() and GetUpdatedTestExpectations(). Move FindModifiedLayoutTests().
(TestFactory): Set skipBuildIfNoResult.
(iOSTestsFactory): Set findModifiedLayoutTests.
(macOSWK1Factory): Ditto.
(macOSWK2Factory): Ditto.
(GTKTestsFactory): Ditto.
(WPETestsFactory): Ditto.

* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps):

* Tools/CISupport/ews-build/steps.py:
(GetTestExpectationsBaseline): Print test expectations before applying changes.
(GetTestExpectationsBaseline.run):
(GetUpdatedTestExpectations): Print test expectations again after applying changes.
(GetUpdatedTestExpectations.run):
(FindModifiedLayoutTests): Parse diff between two expectations and check patch for LayoutTest changes.
(FindModifiedLayoutTests.run):
(FindModifiedLayoutTests.getResultSummary):
(FindModifiedLayoutTests.start): Deleted.

* Tools/CISupport/ews-build/steps_unittest.py:
(TestGetTestExpectationsBaseline): Added.
(TestGetUpdatedTestExpectations): Added.
(TestFindModifiedLayoutTests):
(TestFindModifiedLayoutTests.test_non_accessible_patch): Test for non-accessible patch/PR contents.

Canonical link: https://commits.webkit.org/277546@main

webkit-commit-queue commented:
Committed 277546@main (e530f8ef12ba): <https://commits.webkit.org/277546@main>

Reviewed commits have been landed. Closing PR #26478 and removing active labels.
